cli/src/semgrep/fork_subprocess.py

```diff
@@ -0,0 +1,225 @@
+# fork_subprocess.py
+"""
+Define the fork_subprocess.Process class, which uses the
+'multiprocessing' library to fork a child process, then expose its
+stdout and stderr as 'asyncio.StreamReaders' in the parent.
+"""
+import asyncio  # StreamReader, gather, run
+import multiprocessing  # Process
+import os  # fdopen, pipe, dup2, etc.
+from typing import Callable
+
+
+# This is a copy of asyncio.subprocess._DEFAULT_LIMIT.
+DEFAULT_LIMIT = 2**16  # 64 KiB
+
+
+# Related SO Q+A:
+#   https://stackoverflow.com/questions/52089869/how-to-create-asyncio-stream-reader-writer-for-stdin-stdout
+#   https://stackoverflow.com/questions/64303607/python-asyncio-how-to-read-stdin-and-write-to-stdout
+async def reader_for_pipe_fd(fd: int, limit: int) -> asyncio.StreamReader:
+    """
+    Return a StreamReader for pipe file descriptor 'fd'.
+
+    The 'limit' defines the maximum size of a single 'read' operation to
+    avoid the possibility of deadlock.
+    """
+    # Get the running loop (which must exist), and pass it to the new
+    # objects so everyone is using the same loop.
+    loop = asyncio.get_running_loop()
+
+    # Build the reader to return.  StreamReader exposes an asynchronous
+    # interface for reading.  It requires that something synchronously
+    # feed it data that it can provide to the client.
+    #
+    # I do not know what 'limit' really represents (the documentation is
+    # vague).  It is used by some sketchy-looking logic inside
+    # StreamReader; I suspect a latent bug.  This function's
+    # documentation contains my empirical observation.
+    reader = asyncio.StreamReader(loop=loop, limit=limit)
+
+    # Wrap a Python file object around 'fd'.  It does not own 'fd';
+    # the client will continue to own it.  This file object does not get
+    # explicitly closed.
+    #
+    # Internally, 'connect_read_pipe' basically just calls 'fileno()' to
+    # get the descriptor back out, and insists it is a pipe or socket.
+    # It does not work with arbitrary files.
+    fd_file = os.fdopen(fd, "rb", closefd=False)
+
+    # Add 'fd_file' to the set of descriptors monitored by 'loop' (and
+    # remove it upon EOF).
+    #
+    # Although 'connect_read_pipe' is 'async', it does not really block;
+    # it will succeed or fail immediately, although notification takes a
+    # detour through the 'call_soon' queue.
+    await loop.connect_read_pipe(
```

Comment:
is there a maximum capacity for the read buffer we should worry about?

Reply:
Not for `connect_read_pipe` itself, but yes once the pipe starts being used.  This is handled in `core_runner.py` by passing `limit=INPUT_BUFFER_LIMIT` in a manner parallel to the pathway that uses `asyncio.create_subprocess_exec`.  There is an existing test (`test/e2e/test_performance.py`) that fails if the limit is not set properly.